### DISCUSSION TITLE
How to get costs associated with an API call / user?

### DISCUSSION THREAD

**Discussion by punkpeye** [langfuse/langfuse#3466] [Support] [answered]
Is there a way to get LLM request cost back in my application?

I am founder of https://glama.ai and my use case is to track costs associated with individual users.

I would ideally want a webhook or something like that that would tell me how much a request cost.

Update: ended up building something that does this https://glama.ai/blog/2024-09-25-chat-cost-calculator-api

**Comment by marcklingen** [comment, marked as answer]
Hi @punkpeye,

Depending on your needs you can use:
- Timeseries with sum of daily cost/usage per user via [daily-metrics api](https://langfuse.com/docs/analytics/daily-metrics-api) filtered by user. Commonly used for billing, usage limits, user-facing dashboards and internal reporting.
- List of last n traces and their costs via [`GET /traces`](https://langfuse.com/docs/query-traces) (sdks: `fetch_traces`/`fetchTraces`) filtered by userId. Note that there is no write/read consistency as traces are persisted asynchronously, thus it might take a couple of seconds (usually) and up to a few minutes (sudden very high peak in usage).

Does this solve for your goal?

I think webhooks would be super useful here but are not yet available. You can upvote [this idea thread](https://github.com/orgs/langfuse/discussions/1033) and add your thoughts on why this would be useful for you to help prioritize this on our roadmap.

**Reply by punkpeye** [reply]
Yeah, webhooks would definitely be useful. Furthermore, I would happily pay just for API access that allows me to calculate such costs. It is not something that I am personally eager to maintain.

**Reply by marcklingen** [reply]
Thanks for the upvote on webhooks.

Did you check the metrics api which can return aggregated cost data for each user? Is this what you’re looking for?

**Reply by punkpeye** [reply]
I would it need it per request since that's how I keep track of costs internally.

**Reply by marcklingen** [reply]
Makes sense. Then you can fetch the last requests for each user or for all users (which contains the userId) via the api mentioned above (option 2).

**Reply by punkpeye** [reply]
Is there a way to assign a request ID at the time of making it, and then fetch info about that request later?

**Reply by marcklingen** [reply]
yes, you can set your own request id as a trace id (see integration docs) and then fetch the costs for this id later on via fetch_trace (= GET /traces/[id])

**Reply by punkpeye** [reply]
@marcklingen Is the cost guaranteed to be accurate? e.g. I noticed that Gemini charges different token rate depending on the total context length, which other calculators do not take into account.

**Reply by marcklingen** [reply]
No this is not guaranteed with a current limitation that costs for llm calls that use context caching and variable prices (as for example for Gemini) need to be added via api when logging traces to Langfuse. We plan to update cost tracking in Langfuse to support these pricing models more natively. Have a look at the cost tracking docs for further reference.
